File: basic_dsa_one_hundred_ninety_one.java
/**
 * Basic DSA Problems - Copy List with Random Pointer
 * Deep copy a linked list where each node has next and random pointer
 */
import java.util.HashMap;
import java.util.Map;

public class basic_dsa_one_hundred_ninety_one {

    // Problem: Each node has next and random. Return deep copy of the list.

    public static class Node {
        int val;
        Node next, random;
        Node(int val) { this.val = val; }
    }

    /**
     * Map old -> new. First pass create copies; second pass set next and random.
     * Time O(n), Space O(n)
     */
    public static Node copyRandomList(Node head) {
        if (head == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Node cur = head;
        while (cur != null) {
            map.put(cur, new Node(cur.val));
            cur = cur.next;
        }
        cur = head;
        while (cur != null) {
            Node copy = map.get(cur);
            copy.next = map.get(cur.next);
            copy.random = map.get(cur.random);
            cur = cur.next;
        }
        return map.get(head);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Copy List with Random Pointer");
        System.out.println("=".repeat(60));

        Node a = new Node(7);
        Node b = new Node(13);
        Node c = new Node(11);
        Node d = new Node(10);
        Node e = new Node(1);
        a.next = b; b.next = c; c.next = d; d.next = e;
        a.random = null; b.random = a; c.random = e; d.random = c; e.random = a;

        Node copy = copyRandomList(a);
        System.out.println("Copy head val -> " + (copy != null ? copy.val : "null"));
    }
}

File: basic_dsa_one_hundred_ninety_one.py
"""
Basic DSA - Rotate Array
Given an integer array nums, rotate the array to the right by k steps, where k is
non-negative.
"""


def rotate(nums, k):
    """
    Reverse parts:
    - reverse whole array
    - reverse first k
    - reverse rest
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    n = len(nums)
    if n == 0:
        return nums
    k %= n

    def rev(i, j):
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1

    rev(0, n - 1)
    rev(0, k - 1)
    rev(k, n - 1)
    return nums


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Rotate Array")
    print("=" * 60)

    a = [1, 2, 3, 4, 5, 6, 7]
    rotate(a, 3)
    print(a)  # [5, 6, 7, 1, 2, 3, 4]

    b = [-1, -100, 3, 99]
    rotate(b, 2)
    print(b)  # [3, 99, -1, -100]
